packages/core/contracts/oracle/implementation/OptimisticOracle.sol

```diff
@@ -0,0 +1,368 @@
+        Request storage request = _getRequest(requester, identifier, timestamp);
+        request.proposer = proposer;
+        request.proposedPrice = proposedPrice;
+        // If a custom liveness has been set
+        request.expirationTime = getCurrentTime().add(
+            request.customLiveness != 0 ? request.customLiveness : defaultLiveness
+        );
+
+        uint256 totalBond = request.bond.add(request.finalFee);
+        if (totalBond > 0) {
+            request.currency.safeTransferFrom(msg.sender, address(this), totalBond);
+        }
+
+        // Event.
+        emit ProposePrice(requester, proposer, identifier, timestamp, proposedPrice);
+
+        // Callback.
+        try OptimisticRequester(requester).priceProposed(identifier, timestamp) {} catch {}
+    }
+
+    function proposePrice(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 proposedPrice
+    ) external {
+        // Note: re-entrancy guard is done in the inner call.
+        proposePriceFor(msg.sender, requester, identifier, timestamp, proposedPrice);
+    }
+
+    function disputePriceFor(
+        address disputer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp
+    ) public nonReentrant() {
+        require(getState(requester, identifier, timestamp) == State.Proposed, "disputePriceFor: Proposed");
+        Request storage request = _getRequest(requester, identifier, timestamp);
+        request.disputer = disputer;
+
+        uint256 finalFee = request.finalFee;
+        uint256 totalBond = request.bond.add(finalFee);
+        if (totalBond > 0) {
+            request.currency.safeTransferFrom(msg.sender, address(this), totalBond);
+        }
+
+        StoreInterface store = _getStore();
+        if (finalFee > 0) {
+            request.currency.safeIncreaseAllowance(address(store), finalFee);
+            _getStore().payOracleFeesErc20(address(request.currency), FixedPoint.Unsigned(finalFee));
+        }
+
+        _getOracle().requestPrice(identifier, timestamp);
+
+        // Compute refund.
+        uint256 refund = 0;
+        if (request.reward > 0 && request.refundOnDispute) {
+            refund = request.reward;
+            request.reward = 0;
+            request.currency.safeTransfer(requester, refund);
+        }
+
+        // Event.
+        emit DisputePrice(requester, request.proposer, disputer, identifier, timestamp);
+
+        // Callback.
+        try OptimisticRequester(requester).priceDisputed(identifier, timestamp, refund) {} catch {}
+    }
+
+    function disputePrice(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp
+    ) external {
+        // Note: re-entrancy guard is done in the inner call.
+        disputePriceFor(msg.sender, requester, identifier, timestamp);
+    }
+
+    function getPrice(bytes32 identifier, uint256 timestamp) external nonReentrant() returns (int256) {
```

Comment:
Would we make this an `onlyRegisteredContract` method?

Reply:
This is a really good point.

I _don't_ think we do. Would be great to hear your thoughts here: the "master plan" is for this optimistic oracle to effectively allow callers to be trial users of the DVM (this is something @hlambur has been pushing for to improve UX). If the caller is paying their fees fairly then the optimistic oracle should work fine. If the caller is not paying fees, the DVM voters can choose to respond with an invalid value (INT_MIN, I was thinking). This is something we don't have to 100% figure out now since a the logic is basically off-chain, but it would be nice to develop this in a way that's compatible with the DVM trial idea.

What are your thoughts?

Reply:
I think it should be `onlyRegisteredContract` until we figure out the economic consequences of opening up DVM state to everyone. I think this gets into parasitic DVM usage.

Reply:
My only worry with doing that is having it audited with that restriction and then needing another audit before we can open it up.

Reply:
Hm I'm just not sure this would be caught by a security audit and it doesn't seem to increase the "code" attack vector. It just might increase the economic attack vector, I think.